File: widgetssdk/src/main/java/com/glia/widgets/core/screensharing/MediaProjectionService.java
package com.glia.widgets.core.screensharing;

import android.app.Notification;
import android.app.Service;
import android.content.Intent;
import android.content.pm.ServiceInfo;
import android.os.Build;
import android.os.IBinder;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;

import com.glia.androidsdk.Engagement;
import com.glia.androidsdk.Glia;
import com.glia.androidsdk.omnibrowse.Omnibrowse;
import com.glia.androidsdk.omnibrowse.OmnibrowseEngagement;
import com.glia.androidsdk.omnicore.OmnicoreEngagement;
import com.glia.widgets.core.notification.NotificationFactory;

import java.util.function.Consumer;

/**
 * Apps targeting SDK version 29 or later require for screen-sharing a running foreground service.
 * This service requires declaration in the Manifest file.
 */
@RequiresApi(api = Build.VERSION_CODES.O)
public class MediaProjectionService extends Service {
    private static final String TAG = "MediaProjectionService";
    private static final int SERVICE_ID = 123;
    private final Consumer<OmnicoreEngagement> omnicoreEngagementStartListener = (engagement) ->
        engagement.on(Engagement.Events.END, (Runnable) this::stopSelf);
    private final Consumer<OmnibrowseEngagement> omnibrowseEngagementStartListener = (engagement) ->
        engagement.on(Engagement.Events.END, (Runnable) this::stopSelf);

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        setupAsForegroundService();
    }

    public void setupAsForegroundService() {
        // Register this service as a foreground service.
        Notification screenSharingNotification = NotificationFactory.createScreenSharingNotification(this);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            startForeground(SERVICE_ID, screenSharingNotification, ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PROJECTION);
        } else {
            startForeground(SERVICE_ID, screenSharingNotification);
        }
    }

    @Override
    public void onTaskRemoved(Intent rootIntent) {
        stopSelf();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        String action = null;
        if (intent != null) {
            action = intent.getAction();
        }
        if (Actions.START.equals(action)) {
            setupListeners();
        }

        return Service.START_STICKY;
    }

    public void setupListeners() {
        Glia.on(Glia.Events.ENGAGEMENT, omnicoreEngagementStartListener);
        Glia.omnibrowse.on(Omnibrowse.Events.ENGAGEMENT, omnibrowseEngagementStartListener);
    }

    public void removeListeners() {
        if (!Glia.isInitialized()) return;
        Glia.off(Glia.Events.ENGAGEMENT, omnicoreEngagementStartListener);
        Glia.omnibrowse.off(Omnibrowse.Events.ENGAGEMENT, omnibrowseEngagementStartListener);
    }

    @Override
    public void onDestroy() {
        removeListeners();
        super.onDestroy();
    }

    public interface Actions {
        String START = "EngagementMonitoringService:Start";
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/core/screensharing/MediaProjectionService.kt
package com.glia.widgets.core.screensharing

import android.app.Service
import android.content.Intent
import android.content.pm.ServiceInfo
import android.os.Build
import android.os.IBinder
import androidx.annotation.RequiresApi
import com.glia.widgets.internal.notification.NotificationFactory.createScreenSharingNotification
import com.glia.widgets.di.Dependencies
import com.glia.widgets.engagement.State
import com.glia.widgets.engagement.domain.EngagementStateUseCase
import com.glia.widgets.engagement.domain.InformThatReadyToShareScreenUseCase
import io.reactivex.rxjava3.disposables.CompositeDisposable

private const val SERVICE_ID = 123
internal const val MEDIA_PROJECTION_SERVICE_ACTION_START = "com.glia.widgets.core.screensharing.MediaProjectionService:Start"

/**
 * Glia internal class.
 *
 * It will be automatically added to the integrator's manifest file by the manifest merger during compilation.
 *
 * Apps targeting SDK version 29 or later require a running foreground service for screen sharing.
 */
@RequiresApi(api = Build.VERSION_CODES.O)
internal class MediaProjectionService : Service() {
    private val compositeDisposable: CompositeDisposable by lazy { CompositeDisposable() }
    private val engagementStateUseCase: EngagementStateUseCase by lazy { Dependencies.useCaseFactory.engagementStateUseCase }
    private val informThatReadyToShareScreenUseCase: InformThatReadyToShareScreenUseCase by lazy { Dependencies.useCaseFactory.readyToShareScreenUseCase }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onCreate() {
        super.onCreate()
        setupAsForegroundService()
    }

    private fun setupAsForegroundService() {
        // Register this service as a foreground service.
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            startForeground(SERVICE_ID, createScreenSharingNotification(this), ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PROJECTION)
        } else {
            startForeground(SERVICE_ID, createScreenSharingNotification(this))
        }
    }

    override fun onTaskRemoved(rootIntent: Intent) {
        stopSelf()
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        if (intent.action == null) {
            return START_NOT_STICKY
        }

        if (MEDIA_PROJECTION_SERVICE_ACTION_START == intent.action) {
            informThatReadyToShareScreenUseCase()

            engagementStateUseCase()
                .filter { it is State.EngagementEnded }
                .subscribe { stopSelf() }
                .also(compositeDisposable::add)
        }
        return START_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()
        compositeDisposable.clear()
    }
}
